### PR TITLE
Abstract FFT interface

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/dask/dask/issues/2092

Provides wrappers for things that implement the standard FFT interface like that of NumPy. Recasts the existing FFT interface with these wrappers. Also adds some tests that exercise this wrapping with NumPy and SciPy.

### CODE REVIEW

File: dask/array/fft.py [hunk 1]
```diff
@@ -86,19 +89,73 @@ def _ihfft_out_chunks(a, n, axis):
...
+    return _fft_wrap(np.float_, _hfft_out_chunks)(fft_func)
+
+
+def ihfft_wrapper(fft_func):
+    """ wrapper of 1D hermitian complex inverse FFT functions
+
+    takes a function that behaves like ``numpy.fft.ihfft``.
+    """
+
+    return _fft_wrap(np.complex_, _ihfft_out_chunks)(fft_func)
```
Comment (jcrist) [inline dask/array/fft.py]: Perhaps instead of defining a new wrapper for each type, define a single function that dispatches on the type? Something like:

```python
ifft = fft_wrapper(npfft.ifft, kind='ifft')  # Could use a better keyword name
hfft = fft_wrapper(npfft.hfft, kind='hfft')
...
```

File: dask/array/tests/test_fft.py [hunk 2]
```diff
@@ -108,3 +113,41 @@ def test_fft_consistent_names():
...
+def test_wrap_scipy():
+    spfft = pytest.importorskip("scipy.fftpack")
+
+    fft = fft_wrapper(spfft.fft)
+    assert_eq(fft(darr), spfft.fft(nparr))
+    assert_eq(fft(darr2, axis=0), spfft.fft(nparr, axis=0))
+
+    ifft = ifft_wrapper(spfft.ifft)
+    assert_eq(ifft(darr), spfft.ifft(nparr))
+    assert_eq(ifft(darr2, axis=0), spfft.ifft(nparr, axis=0))
```
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: As SciPy doesn't have `hfft` or `ihfft`, we can't wrap those.

Also they implement `rfft` and `irfft`, but how they handle the data is a bit strange. Instead of returning or taking complex values all of these functions return real values that split the imaginary part from the real part. So wrapping them would be kind of a pain as we would have to massage the data back into the expected form in the wrapper.

ref: https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.fftpack.rfft.html

File: dask/array/fft.py (hunk 1 above)
Comment (jakirkham) [inline dask/array/fft.py]: I'm 50/50 on it. What's the disadvantage of the current form?

**Comment by jakirkham** [comment]
Given you wrote the original implementation, @cowlicks, it would be good to hear your feedback on this.

File: dask/array/fft.py (hunk 1 above)
Comment (jcrist) [inline dask/array/fft.py]: Just a lot of functions. I'm slightly in favor of a smaller public interface and a more concise implementation. Not a big deal though, current way is fine as well.

Comment (jakirkham) [inline dask/array/fft.py]: Well, I'm planning to add the 2-D and N-D versions of these too. So as long as we don't think that this will make the function too long with such a big branch that is fine. Though some linting tools may balk at this.

Comment (mrocklin) [inline dask/array/fft.py]: I'm also in favor of fewer functions here if we can manage it.  It makes it easier to maintain and change in the future if need arises.

File: dask/array/tests/test_fft.py [hunk 3]
```diff
@@ -108,3 +113,41 @@ def test_fft_consistent_names():
     assert same_keys(fft(darr, 5), fft(darr, 5))
     assert same_keys(fft(darr2, 5, axis=0), fft(darr2, 5, axis=0))
     assert not same_keys(fft(darr, 5), fft(darr, 13))
+
+
+def test_wrap_numpy():
```
Comment (mrocklin) [inline dask/array/tests/test_fft.py]: Can we refactor this to use `pytest.mark.parametrize`?  This might make it easier to extend in the future.

```python
@pytest.mark.parametrize('module', [np.fft, scipy.fft])
@pytest.mark.parametrize('funcname', ['fft', 'ifft', 'rfft', ...])
def test_wrapper(module, funcname):
    try:
        func = getattr(module, funcname)
    except AttributeError:
        pytest.mark.skip()
    dfft = fft_wrapper(module, funcname)
    assert_eq(fft(nparr), dfft(darr))
```

Comment (mrocklin) [inline dask/array/tests/test_fft.py]: Also I have a slight preference to use names like `x` and `dx` rather than `nparr`.  These names were probably from before, but this refactor might be a nice chance to rename them.

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: > Can we refactor this to use `pytest.mark.parametrize`?

Probably to some extent. Though I don't think we can lump everything into one test. Please see this [note]( https://github.com/dask/dask/pull/2093#discussion_r106466514 ).

Comment (mrocklin) [inline dask/array/tests/test_fft.py]: Is the PyFFTW library consistent with the numpy interface?

File: dask/array/tests/test_fft.py (hunk 2 above)
Comment (mrocklin) [inline dask/array/tests/test_fft.py]: Perhaps it doesn't matter as long as `fft_wrap` ensures that the dask.array version is consistent with whatever function it is passed.  We don't need to test against numpy, we just need to test against the function that we were given.

File: dask/array/fft.py [hunk 4]
```diff
@@ -24,21 +24,24 @@
...
+            if len(a.chunks[axis]) != 1:
+                raise ValueError(chunk_error % (axis, a.chunks[axis]))
 
-        chunks = out_chunk_fn(a, n, axis)
+            chunks = out_chunk_fn(a, n, axis)
 
-        return map_blocks(partial(fft_func, n=n, axis=axis), a, dtype=dtype,
-                          chunks=chunks)
+            return map_blocks(partial(fft_func, n=n, axis=axis), a, dtype=dtype,
+                              chunks=chunks)
```
Comment (mrocklin) [inline dask/array/fft.py]: It is useful to avoid partial.  map_blocks can accept keyword arguments, I recommend the following:

    map_blocks(fft_func, a, n=n, axis=axis, dtype=dtype, chunks=chunks)

File: dask/array/tests/test_fft.py (hunk 3 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: These [other variables]( https://github.com/dask/dask/blob/175ae5a91ac09940c1065ea007a6a93a95d39207/dask/array/tests/test_fft.py#L20-L22 ) were already defined and use in the other tests. I used them to be consistent and to ensure the wrappers were doing the correct thing. I would like to keep this change focused on the task at hand.

Comment (mrocklin) [inline dask/array/tests/test_fft.py]: Yeah, I wasn't blaming you for style things.  Just a request.  I'm fine keeping things focused if you prefer.

File: dask/array/fft.py (hunk 4 above)
Comment (jakirkham) [inline dask/array/fft.py]: This line was already here. I have merely indented it. While there may be improvements that can be made to this function, I'd like to keep this change compact.

File: dask/array/fft.py (hunk 1 above)
Comment (jakirkham) [inline dask/array/fft.py]: Ok.

File: dask/array/tests/test_fft.py (hunk 3 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: > Yeah, I wasn't blaming you for style things. Just a request. I'm fine keeping things focused if you prefer.

No worries. I didn't think you were. Maybe we should document a set of things like this that we would like to see improved in a separate issue. As I'll likely be back to muck with this code, I can try to pick it up at some point.

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: > Is the PyFFTW library consistent with the numpy interface?

Yes, they provide a numpy interface that they try to keep up-to-date. Here's the [docs]( https://hgomersall.github.io/pyFFTW/pyfftw/interfaces/numpy_fft.html#module-pyfftw.interfaces.numpy_fft ) for more details.

Was debating adding optional tests for it, but was unclear if we want to go that route. One thing to consider is FFTW is GPL (or some license fee to remove GPL) while pyFFTW is BSD. Maybe it would be ok to an optional test. I'll defer to you though.

Comment (mrocklin) [inline dask/array/tests/test_fft.py]: I don't know what the legality is when you're testing against other code.

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: IANAL. That said, my understanding was as long as the GPL dependencies are optional and are not shipped with the distribution, this should not effect licensing. This SO [answer]( http://softwareengineering.stackexchange.com/a/175905 ) seems pertinent.

Note: FFTW is GPL 2.

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: That being said, if you would rather steer clear, I'm ok with that. It is part of the reason I designed these to be user accessible wrappers as opposed to discussing optional FFTW support. 😄

**Comment by cowlicks** [comment]
👋 @jakirkham I haven't touched dask in a while, so I'll defer to @mrocklin and @jcrist cheers!

File: dask/array/fft.py (hunk 1 above)
Comment (jakirkham) [inline dask/array/fft.py]: Done.

File: dask/array/tests/test_fft.py (hunk 3 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Have parameterized the tests. Have not done anything with pyFFTW. We can save that for a follow-up PR.

File: dask/array/tests/test_fft.py (hunk 2 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: So I think it is worth testing that the computation is correct. Otherwise we have no idea whether the wrapper is behaving correctly from the test side.

**Comment by jakirkham** [comment]
Ok, I think I have addressed the concerns raised with this patch. Would appreciate another review.

File: dask/array/fft.py [hunk 5]
```diff
@@ -86,19 +89,29 @@ def _ihfft_out_chunks(a, n, axis):
...
+    elif kind == "ifft":
+        return _fft_wrap(np.complex_, _fft_out_chunks)(fft_func)
+    elif kind == "rfft":
+        return _fft_wrap(np.complex_, _rfft_out_chunks)(fft_func)
+    elif kind == "irfft":
+        return _fft_wrap(np.float_, _irfft_out_chunks)(fft_func)
+    elif kind == "hfft":
+        return _fft_wrap(np.float_, _hfft_out_chunks)(fft_func)
+    elif kind == "ihfft":
+        return _fft_wrap(np.complex_, _ihfft_out_chunks)(fft_func)
```
Comment (jcrist) [inline dask/array/fft.py]: Would be nice to raise a `ValueError` in case the kind is unknown.

File: dask/array/tests/test_fft.py [hunk 6]
```diff
@@ -108,3 +112,25 @@ def test_fft_consistent_names():
...
+
+
+@pytest.mark.parametrize("modname", ["numpy.fft", "scipy.fftpack"])
+@pytest.mark.parametrize("funcname", ["fft", "ifft", "hfft", "ihfft"])
+def test_wrap_ffts(modname, funcname):
+    fft_mod = pytest.importorskip(modname)
+    try:
+        func = getattr(fft_mod, funcname)
+    except AttributeError:
+        raise SkipTest()
```
Comment (jcrist) [inline dask/array/tests/test_fft.py]: Use `pytest.skip('fft method missing')  # or a better message` here instead of raising `SkipTest`

**Comment by jcrist** [comment]
Other than my comments above, this looks good to me. Thanks for the fix @jakirkham.

File: dask/array/fft.py [hunk 7]
```diff
@@ -24,21 +24,24 @@
     """
 
 
-def _fft_wrap(fft_func, dtype, out_chunk_fn):
-    def func(a, n=None, axis=-1):
-        if len(a.chunks[axis]) != 1:
-            raise ValueError(chunk_error % (axis, a.chunks[axis]))
+def _fft_wrap(dtype, out_chunk_fn):
```
Comment (mrocklin) [inline dask/array/fft.py]: Can we auto-detect the dtype by calling the function on a length zero array?  Actually, we may be able to just ignore the dtype argument altogether and let `map_blocks` take care of this (it will call the function on a length one array of the input array's dtype?

Comment (jcrist) [inline dask/array/fft.py]: If we know the dtype ahead of time, I don't see a good reason not to provide that information statically if it's cheap to do.

File: dask/array/fft.py (hunk 5 above)
Comment (mrocklin) [inline dask/array/fft.py]: In the spirit of keeping code as data when possible, I recommend creating a dict like 

```python
{'fft': _fft_out_chunks,
 'ifft': _fft_out_chunk, # also should this be ifft?
 ....
}
```

File: dask/array/fft.py (hunk 7 above)
Comment (mrocklin) [inline dask/array/fft.py]: OK

File: dask/array/tests/test_fft.py (hunk 6 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Ok, not super familiar with `pytest`. So thanks for the pointer. Will update.

File: dask/array/fft.py (hunk 5 above)
Comment (jakirkham) [inline dask/array/fft.py]: > I recommend creating a dict like

We could do that. So a different `dict` for type too?

File: dask/array/tests/test_fft.py (hunk 6 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Done.

File: dask/array/fft.py (hunk 5 above)
Comment (jakirkham) [inline dask/array/fft.py]: > Would be nice to raise a `ValueError` in case the kind is unknown.

Handled and added a test.

Comment (jakirkham) [inline dask/array/fft.py]: > I recommend creating a dict

Also done. Made a `dict` with the parameters so they can be passed as kwargs.

File: dask/array/fft.py (hunk 7 above)
Comment (jakirkham) [inline dask/array/fft.py]: I do like your idea about detecting type better, @mrocklin. Let me give that some more thought.

**Comment by jakirkham** [comment]
Thanks for the follow-up comments. I think I have addressed everything raised. Would appreciate another look.

File: dask/array/fft.py (hunk 7 above)
Comment (jakirkham) [inline dask/array/fft.py]: In the past, I have not seen a great way to go from `float*` to `complex*` that preserves precision (e.g. `float32` to `complex64`). So have wound up handling this [myself]( https://github.com/nanshe-org/nanshe/blob/4998fb87f373d951159c49352ca6cdf28c0a8c62/nanshe/imp/registration.py#L186-L192 ). I think from `complex*` to `float*` this should be easier (e.g. take `real` of `0` or similar), but the other way around ends up being kind of ugly.

File: dask/array/fft.py [hunk 8]
```diff
@@ -86,19 +89,61 @@ def _ihfft_out_chunks(a, n, axis):
...
+        },
+        "hfft": {
+            "dtype": np.float_,
+            "out_chunk_fn" : _hfft_out_chunks
+        },
+        "ihfft": {
+            "dtype": np.complex_,
+            "out_chunk_fn" : _ihfft_out_chunks
+        },
+    }
```
Comment (mrocklin) [inline dask/array/fft.py]: The intent behind the data comment is to keep this data outside of functions.  That way it can be used by other functions as well.

```python
dtypes = {'fft: ...}
funcs = {'fft': ...

def func(name, ...):
    dtype = dtypes[name]
    chunk_func = funcs[name]
```

File: dask/array/tests/test_fft.py [hunk 9]
```diff
@@ -108,3 +110,32 @@ def test_fft_consistent_names():
...
+        func = getattr(fft_mod, funcname)
+    except AttributeError:
+        pytest.skip("`%s` missing function `%s`." % (modname, funcname))
+
+    if modname == "scipy.fftpack" and "rfft" in funcname:
+        pytest.skip(
+            "`%s` function `%s` doesn't support the NumPy API." % (
+                modname, funcname
+            )
+        )
```
Comment (mrocklin) [inline dask/array/tests/test_fft.py]: Would the AttributeError check above nt cover this?

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: No because as noted in this [comment]( https://github.com/dask/dask/pull/2093#discussion_r106466514 ) `rfft` and  `irfft` do exist in SciPy, but they format data in a totally different way than NumPy's `rfft` and  `irfft`. So we have to massage it into a usable format if we want to test wrapping this function. Not sure it is worth it IMHO.

**Comment by mrocklin** [comment]
@jakirkham I've pushed a few commits here.  If you have a chance can you take a look and say if I'm missing anything important?

File: dask/array/fft.py [hunk 10]
```diff
@@ -86,19 +64,64 @@ def _ihfft_out_chunks(a, n, axis):
...
+        * hfft
+        * ihfft
+
+    Examples
+    --------
+    >>> parallel_fft = fft_wrap(np.fft.fft)
+    >>> parallel_ifft = fft_wrap(np.fft.ifft)
+    """
+    if kind is None:
+        kind = fft_func.__name__
```
Comment (jakirkham) [inline dask/array/fft.py]: Honestly I would think explicit is better than implicit.

File: dask/array/fft.py [hunk 11]
```diff
@@ -86,19 +64,64 @@ def _ihfft_out_chunks(a, n, axis):
     return chunks
 
 
-fft = _fft_wrap(npfft.fft, np.complex_, _fft_out_chunks)
+out_chunk_fns = {'fft': _fft_out_chunks,
+                 'ifft': _fft_out_chunks,
+                 'rfft': _rfft_out_chunks,
+                 'irfft': _irfft_out_chunks,
+                 'hfft': _hfft_out_chunks,
+                 'ihfft': _ihfft_out_chunks}
```
Comment (jakirkham) [inline dask/array/fft.py]: So I'm not sure it is a good idea to make `out_chunk_fns` as part of the public API. Particularly when what it is mapping are functions that are not part of the public API.

**Comment by jakirkham** [comment]
I'm unclear on how you are handling the SciPy `*rfft` functions. They take/provide different shapes and types than their NumPy equivalents. However, it looks like we are using the same chunking functions. Is that correct? There is a lot in here now. So I might be missing something.

File: dask/array/fft.py (hunk 10 above)
Comment (mrocklin) [inline dask/array/fft.py]: Yes, but concise is also better than needlessly repetitive.  Tradeoffs exist.

File: dask/array/fft.py (hunk 11 above)
Comment (mrocklin) [inline dask/array/fft.py]: I'd be happy to prepend with an underscore if that would calm this concern.

**Comment by mrocklin** [comment]
> I'm unclear on how you are handling the SciPy *rfft functions. They take/provide different shapes and types than their NumPy equivalents. However, it looks like we are using the same chunking functions. Is that correct? There is a lot in here now. So I might be missing something.

Can you suggest a test that would expose this difference?

The typing issue is handled by just running the function on some empty data if the dtype is not provided.  I haven't taken a look at the chunking code; it's possible that we're missing something there.  Assert_eq is checking that expected shape matches computed shape though, so at least for the arrays in the parametrized tests we're OK so far.

File: dask/array/fft.py (hunk 10 above)
Comment (mrocklin) [inline dask/array/fft.py]: Or, more verbosely, I'm against providing both `np.fft.fft` and the name `'fft'`.  This seems needlessly verbose.  Do the names ever differ from the operation in any of numpy/scipy/pyfftw?  If someone came with another set of functions that don't match then they can always provide the "kind" explicitly.

**Comment by mrocklin** [comment]
I've changed out_chunk_fns to be private.  I would like to merge this soon.

File: dask/array/fft.py (hunk 10 above)
Comment (jakirkham) [inline dask/array/fft.py]: Just sharing an extremely cautious opinion. Feel free to go ahead inspite of it.

> Do the names ever differ from the operation in any of numpy/scipy/pyfftw?

I don't think so.

Though how the `*rfft` functions behave in scipy (and pyfftw's scipy interface) are different that numpy's (and pyfftw's numpy interface). So handling that case correctly is more important than this.

File: dask/array/fft.py (hunk 11 above)
Comment (jakirkham) [inline dask/array/fft.py]: Yep, that would be fine for me. Thanks.

File: dask/array/fft.py (hunk 10 above)
Comment (mrocklin) [inline dask/array/fft.py]: > Though how the *rfft functions behave in scipy (and pyfftw's scipy interface) are different that numpy's (and pyfftw's numpy interface). So handling that case correctly is more important than this.

Understood.  My goal is for `fft_wrap` to mimic the behavior of whatever function it is given, not to match numpy's behavior.  So far we seem to achieve this goal.

**Comment by jakirkham** [comment]
Thanks for helping out with this @mrocklin. Tried to do a short review yesterday, but didn't have lots of time. So it got cut short.

> > I'm unclear on how you are handling the SciPy *rfft functions. They take/provide different shapes and types than their NumPy equivalents. However, it looks like we are using the same chunking functions. Is that correct? There is a lot in here now. So I might be missing something.

> Can you suggest a test that would expose this difference?

Below is a brief example that demonstrates this difference. This borrow from the [docs]( https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.fftpack.rfft.html ) and tweaked slightly. Note that SciPy alternates between floating values that represent real and imaginary components (with an optional final imaginary number for an odd array length) of individual values. NumPy OTOH simply gives us the complex values. 

```python
In [1]: import numpy

In [2]: import numpy.fft

In [3]: import numpy.fft as npfft

In [4]: import scipy

In [5]: import scipy.fftpack

In [6]: import scipy.fftpack as spfft

In [7]: a = numpy.array([9.0, -9.0, 1.0, 3.0])

In [8]: b = numpy.array([-9.0, 1.0, 3.0])

In [9]: npfft.rfft(a)
Out[9]: array([  4. +0.j,   8.+12.j,  16. +0.j])

In [10]: spfft.rfft(a)
Out[10]: array([  4.,   8.,  12.,  16.])

In [11]: npfft.rfft(b)
Out[11]: array([ -5.+0.j        , -11.+1.73205081j])

In [12]: spfft.rfft(b)
Out[12]: array([ -5.        , -11.        ,   1.73205081])
```

File: dask/array/fft.py (hunk 10 above)
Comment (jakirkham) [inline dask/array/fft.py]: Indeed. Playing around with it locally seems ok. Maybe we should test a couple of `n`s for good measure.

**Comment by jakirkham** [comment]
Looking more closely at what SciPy is doing, it seems that with `rfft` one will always get back an array of length `n`. So even though its result type is different and how it structures data is different, we don't actually have to do anything differently except use a different storage type (a real one instead of a complex one).

**Comment by mrocklin** [comment]
That sounds perfect and very easy to support on our end.

Thank you for doing the research there.

File: dask/array/fft.py [hunk 12]
```diff
@@ -86,19 +64,64 @@ def _ihfft_out_chunks(a, n, axis):
...
+    """
+    if kind is None:
+        kind = fft_func.__name__
+    try:
+        out_chunk_fn = _out_chunk_fns[kind]
+    except KeyError:
+        raise ValueError("Given unknown `kind` %s." % kind)
 
+    if dtype is None:
+        dtype = fft_func(np.ones(8)).dtype
```
Comment (jakirkham) [inline dask/array/fft.py]: Would it be ok/reasonable if we delayed this check until `func` is called and leverage `a`'s `dtype`? The main reason I'm thinking about this is if a user provided `float32` array instead of a `float64` one. Then the result would be some single precision type. However, this line would still treat it as double precision. I suppose they can override the `dtype`. Just thinking of catching a few more cases automatically.

Comment (mrocklin) [inline dask/array/fft.py]: That makes sense to me.  I spot checked this on numpy.fft and found that it always returned full precision float and complex dtypes, even when given small data.  However this assumption may not hold in the other modules.  Moving it inside `func` sounds like a good idea to me.

Comment (jakirkham) [inline dask/array/fft.py]: Cool, done. Also added some tests for it.

File: dask/array/tests/test_fft.py [hunk 13]
```diff
@@ -137,6 +137,7 @@ def test_wrap_ffts(modname, funcname, dtype):
 
     wfunc = fft_wrap(func)
     assert wfunc(darrc).dtype == func(nparrc).dtype
+    assert wfunc(darrc).shape == func(nparrc).shape
```
Comment (mrocklin) [inline dask/array/tests/test_fft.py]: The `assert_eq` function already does this (or if it doesn't, then this fix should go there)

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Ok, well I just noticed that this caused a failure as I'm seeing shape mismatches with `*rfft` on SciPy. This actually doesn't surprise me as SciPy's functions preserve the original array's shape (up to `n`). However, NumPy's `*rfft`s don't.

Comment (mrocklin) [inline dask/array/tests/test_fft.py]: OK, so I guess this gets back to your original question about if we should support scipy `rfft` functions, and, if not, how to exclude them.  It seems like the current chunk_fn for rfft is specific to `numpy`.  Do we always ask the user to provide a chunk function of their own?

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: So I looked at the comparison code and it is checking the shapes, but only after computing the dask arrays. The problem is the shape of the dask is wrong, but the shape of the computed result is correct. Can add a PR to compare these too. Though testing that case may prove tricky.

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: PR ( https://github.com/dask/dask/pull/2101 ) should fix `assert_eq` to test the shapes before computation, but it looks like it will expose some existing issues. Not sure if I will have time to fix them all though.

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Ah sorry GitHub hid your previous comment. Well one thing to consider (and this might be the right way to fix it) is SciPy and NumPy provide different return types (real or complex). So the chunking function could take the return type as an argument and infer what to do.

File: dask/array/fft.py [hunk 14]
```diff
@@ -28,31 +30,37 @@ def _fft_out_chunks(a, n, axis):
...
-def _rfft_out_chunks(a, n, axis):
+def _rfft_out_chunks(a, n, axis, rtype):
     if n is None:
         n = a.chunks[axis][0]
     chunks = list(a.chunks)
-    chunks[axis] = (n // 2 + 1,)
+    if issubclass(np.dtype(rtype).type, numbers.Real):
+        chunks[axis] = (n,)
+    else:
+        chunks[axis] = (n // 2 + 1,)
```
Comment (mrocklin) [inline dask/array/fft.py]: Is this code tailored exactly to the numpy and scipy conventions or is there something deeper going on here?

If we're just reverse engineering the scipy conventions into these functions then it might be better to just include these as separate functions.

Comment (jakirkham) [inline dask/array/fft.py]: SciPy's `rfft` always returns a real array. NumPy's `rfft` always returns a complex array. These are documented as such in their respective APIs. So this isn't mysterious. We are just leveraging these known differences.

File: dask/array/fft.py [hunk 15]
```diff
@@ -50,31 +30,37 @@ def _fft_out_chunks(a, n, axis):
...
+def _irfft_out_chunks(a, n, axis, rtype):
     if n is None:
-        n = 2 * (a.chunks[axis][0] - 1)
+        n = a.chunks[axis][0]
     chunks = list(a.chunks)
-    chunks[axis] = (n,)
+    if issubclass(a.dtype.type, numbers.Real):
+        chunks[axis] = (n,)
+    else:
+        chunks[axis] = (2 * (n - 1),)
```
Comment (jakirkham) [inline dask/array/fft.py]: Though if there was one that felt funny, this might be it. We are relying on the type to be the correct type for the `irfft` function we are working with. So if the input array is real, we assume we are dealing with SciPy's `irfft`. However, if the input array is complex, we are assuming we are dealing with NumPy's `irfft`. However, NumPy's `irfft` can take real values in addition to complex ones; whereas, SciPy's `irfft` can only take real values. I'm not sure if I like it much. It might be ok if documented, but it does allow the user to make a potentially painful mistake.

File: dask/array/fft.py (hunk 14 above)
Comment (mrocklin) [inline dask/array/fft.py]: Is it reasonable for someone to come in with a new fft module and it return a complex dtype but not follow scipy's shape scheme?  My question is "are we using dtype to infer which module this is coming from?"  If so then we should probably just condition on the module itself rather than back out which module is being used based on what is happening to the dtype.

Taking a step back, is there a way to better handle the shape business here?  This all seems somewhat convoluted.

Comment (jakirkham) [inline dask/array/fft.py]: I guess, but why? If someone did this, it sort of raises the question of did they understand the purpose of SciPy's API at all. The point was to guarantee real values, keep the same shape, and make it easy to invert later. Making all of the values complex makes the array twice as large, but does not add any value as the imaginary parts will all be zero.

That all said, I can't exactly disagree with you. Trying to handle the SciPy `*rfft` functions is pretty messy and TBH am really not sold on their benefit. They have a few neat tricks, but nothing that really justifies to me why I would want to look at my data in that way instead of an array of complex values.

In short, I'm coming back to the viewpoint that I had when I last looked at this Friday. Namely let's not bother to support the SciPy `*rfft` functions. If someone wants to try later, we have made their job much easier with work done in this PR. Plus we have already opened the door to people wanting the other SciPy FFT functions or using pyFFTW, which is going to help anyone wanting some extra speed to get it easily with Dask.

Comment (mrocklin) [inline dask/array/fft.py]: I don't actually care about this specific situation of complex dtypes.  My question was intended to reinforce this main point: 

> My question is "are we using dtype to infer which module this is coming from?" If so then we should probably just condition on the module itself rather than back out which module is being used based on what is happening to the dtype

I'm ok not supporting scipy functions.  But we need to actively err here, not just avoid testing them.

Generally the goal of this PR appears to be "allow people to come in with their own fft modules".  I guess that the scipy case shows that we can't assume that the "chunk" functions are general.

Comment (jakirkham) [inline dask/array/fft.py]: Ok, let's move towards erroring out if people use SciPy's `*rfft` functions then.

FWIW other people think SciPy's `*rfft` behavior here is weird and are trying to get rid of it.

xref: https://github.com/scipy/scipy/issues/2487

Comment (jakirkham) [inline dask/array/fft.py]: Have added a `ValueError` for those specific SciPy functions and added tests that cover it. Also have reverted all of the special `dtype` handling code.

**Comment by jakirkham** [comment]
Alright, I think this is ready for another look. Please let me know your thoughts.

File: dask/array/tests/test_fft.py [hunk 16]
```diff
@@ -131,25 +131,24 @@ def test_wrap_ffts(modname, funcname, dtype):
...
-    assert_eq(wfunc(darrc, n=len(darrc) - 1),
-              func(nparrc, n=len(darrc) - 1))
-    assert_eq(wfunc(darrc, axis=1, n=darrc.shape[1] - 1),
-              func(nparrc, n=darrc.shape[1] - 1))
-    assert_eq(wfunc(darr2c, axis=0, n=darr2c.shape[0] - 1),
-              func(nparrc, axis=0, n=darr2c.shape[0] - 1))
+    if modname == "scipy.fftpack":
+        if "rfft" in funcname:
+            with pytest.raises(ValueError):
+                wfunc = fft_wrap(func)
```
Comment (mrocklin) [inline dask/array/tests/test_fft.py]: These two if statements should be merged into one line.  Currently you don't test any of the scipy functions except rfft

Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Yep, sorry, have merged the `if`s.

**Comment by jakirkham** [comment]
Made the last tweak and this is now passing. Please let me know if there is anything else.

**Comment by mrocklin** [comment]
Thanks for following through on this @jakirkham .  Merged.

**Comment by jakirkham** [comment]
Thanks for all the feedback and help getting this together.

File: dask/array/tests/test_fft.py (hunk 3 above)
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: FWIW here is a PR ( https://github.com/dask/dask/pull/2105 ) that tests against pyFFTW. Though feel free to accept or decline the patch as you see fit.

File: dask/array/fft.py [hunk 17]
```diff
@@ -86,19 +70,73 @@ def _ihfft_out_chunks(a, n, axis):
...
-hfft = _fft_wrap(npfft.hfft, np.float_, _hfft_out_chunks)
+        chunks = out_chunk_fn(a, n, axis)
+
+        return a.map_blocks(fft_func, n=n, axis=axis, dtype=dtype,
+                            chunks=chunks)
+
+    np_name = fft_func.__name__
+    if fft_func.__doc__ is not None:
+        func.__doc__ = (fft_preamble % (np_name, np_name)) + fft_func.__doc__
+    func.__name__ = np_name
```
Comment (jakirkham) [inline dask/array/fft.py]: Missed fixing the docstring for these wrappers to note that these functions could from different locations. Change was made in PR ( https://github.com/dask/dask/pull/2110 ).

File: dask/array/tests/test_fft.py [hunk 18]
```diff
@@ -24,87 +25,129 @@ def key(k):
...
+    except AttributeError:
+        pytest.skip("`%s` missing function `%s`." % (modname, funcname))
+
+    darrc = darr.astype(dtype)
+    darr2c = darr2.astype(dtype)
+    nparrc = nparr.astype(dtype)
+
+    if modname == "scipy.fftpack" and "rfft" in funcname:
+        with pytest.raises(ValueError):
+            wfunc = fft_wrap(func)
```
Comment (jakirkham) [inline dask/array/tests/test_fft.py]: Forgot that this `wfunc` wasn't doing anything here. So added PR ( https://github.com/dask/dask/pull/2114 ) to drop the variable assignment since it is unneeded.

**Comment by Jeitan** [comment]
I just ran across this and I have a question from a user's perspective, not being familiar with the guts....  but it looks like the scipy FFTs wrapped here going to change to use numpy ffts?  If so I'd like to point a couple things out:

I am often taking FFTs of large datasets, and I have observed that the fft implementations out of scipy.fftpack are significantly better at memory handling than the ones out of numpy.fft.  I've profiled them to very small timesteps and I can catch the numpy FFT with as much as 8 or 16x the memory of the original array (mid-computation), whereas the scipy FFT only uses something like 2 or 4x.  Also, the numpy fft doesn't appear to give back all the memory after it is done.

I would very much like to get all my FFTS from one place (I currently have to use the rfftn from numpy, for example), but I don't want to lose wherever the advantages of scipy.fftpack are coming from.

**Comment by jakirkham** [comment]
I think you are misunderstanding the change, @Jeitan, which is totally understandable as it is in the guts of the library as you say.

Before what happened was Dask wrapped up NumPy's 1-D FFT API and let one use that with Dask Arrays. However there was no easy way to sub in an alternative library like SciPy's FFTPACK or pyFFTW to use with Dask. The functions used to do the wrapping were part of the internal or private API. All we did here was clean up the main wrapping function and expose it in the public API. Dask otherwise still provides the same functions as it did before. This allows you to wrap other FFT functions too as long as they conform to the NumPy API. So you could easily wrap SciPy's FFTPACK (with the exception of `rfft` and `irfft`).* In fact we actually test wrapping these SciPy functions in Dask.

In short, this change should make it easier to explore FFTs from other libraries (like SciPy's) not harder. If you find that to be otherwise, would be curious to hear more about how and why.
<br>

\* SciPy's FFTPACK functions largely conform to the NumPy API with the notable exceptions of `rfft` and `irfft`, which do not. There is a lot of [history]( https://mail.scipy.org/pipermail/numpy-discussion/2006-September/022977.html ) behind this and some discussion about removing/fixing these differences. ( https://github.com/scipy/scipy/issues/2487 ). One might be able to wrap the SciPy `[i]rfft` functions using some [tricks]( http://stackoverflow.com/a/18538345 ) in a wrapper function and then use them with Dask.

**Comment by Jeitan** [comment]
Ahhhhh, I see @jakirkham, thank you for the explanation! Yes I've been following the rfft discussion, to see how that all unfolds (again as an interested user rather than a dev). I appreciate all the hard work that happens here, looking forward to scipy 1.0!
